src/main.c

```diff
@@ -334,11 +391,20 @@ static void check_dfu_mode(void)
     if ( _ota_dfu )
     {
       sd_softdevice_disable();
-    }else
+    }
+    else
     {
-      usb_teardown();
+      channel_teardown();
     }
   }
+
+  // Simply reset, allowing the host MCU to control the behavior on boot via the activation pin.
+  // The host MCU may keep the device in bootloader mode via the DFU activation pin when there are more
+  // binaries to flash (e.g. QSPI flash data)
+  if (PIN_DFU_ACTIVATE_PRESENT && dfu_activate)
+  {
+    NVIC_SystemReset();
```

Comment:
reset immediately here with activation would reduce the modification of existing code. I don't have the set up to test, please double check to see if this commit stills work on your side. Then we can move on with the reviewing (it may take a couple more commits I think).

Reply:
I understand the tear down isn't strictly necessary, but it does no harm either, right?!. :-) I would vote for leaving the code as is, since it's already well tested. I only changed the code that was strictly necessary to keep the changes to the bare minimum.

If you feel very strongly for the change then of course we can test it.

Reply:
please do the testing again, reset here will allow us to keep the rest of main.c unchanged, we could also revert changes to the bootloader.c file as well. I will also make a couple more commits anyway mostly to moving/refactoring changes.

Reply:
I will test this today and will post back with results. Thanks!

Reply:
Unfortunately it didn't work. I can try to find out why later, but from the perspective of the programmer, everything was ok. All the packets were ACKed and it successfully programmed the CPy interpreter and an image for the flash filesystem. But on reset, the nRF stays in the bootloader. 

I'll look deeper into it and check the memory contents and look for other clues as to why it's not working.

Reply:
hmm, that is weird, probably one of the teardown function is indeed required. I will also try to do some check as well

Reply:
I'll review the changes to day to see if I have any suggestions.

Reply:
What I'm seeing is that the app isn't "registered" with the bootloader. This is because of the Reset that you added when `dfu_activate` is `true`. This results in `bootloader_app_start()` never being called.  

That was the reason I separated registering the app from starting it.  The registration is still necessary so that the app is launched when the host MCU restarts the device without the DFU activate pin asserted.

Reply:
I'm still seeing intermittent errors. It is like the application doesn't "stick".

Reply:
I don't have the root cause, but we are able to make this work in cases when the application is the last binary flashed (so qspi filesystem then the app). When the application is flashed first, then the QSPI filesystem, the application isn't saved. I believe the error is on our end (most likely related to timing), because I see the same result when I revert the recent changes to the PR back to my original PR, which I know was working, so something else must have changed.  So I would say this PR is good to merge as is.